### PR TITLE
fix(check-devices): Update --timeout management + add logging management

### PR DESCRIPTION
**Pull request by titom73**
- Fix an issue where timeout was not correctly set.
- Fix timeout defined as an `str` where a `float` was expected
- Add `-log` or `--logging` to set log level for inventory and result_manager

__Default mode__

```bash
check-devices.py -i .personal/avd-lab.yml -c .personal/ceos-catalog.yml --tags arista --username admin --password admin123 --table --timeout 1
╭──────────────────────── Settings ─────────────────────────╮
│ Running check-devices with:                               │
│               - Inventory: .personal/avd-lab.yml          │
│               - Tests catalog: .personal/ceos-catalog.yml │
╰───────────────────────────────────────────────────────────╯
                 All tests results                  
┏━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Device IP ┃ Test Name ┃ Test Status ┃ Message(s) ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
└───────────┴───────────┴─────────────┴────────────┘
```

__With logging enabled__

```bash
❯ check-devices.py -i .personal/avd-lab.yml -c .personal/ceos-catalog.yml --tags arista --username admin --password admin123 --table --timeout 1 --loglevel info
╭──────────────────────── Settings ─────────────────────────╮
│ Running check-devices with:                               │
│               - Inventory: .personal/avd-lab.yml          │
│               - Tests catalog: .personal/ceos-catalog.yml │
╰───────────────────────────────────────────────────────────╯
[14:08:41] WARNING  Service not running on device 10.73.1.254 with:                                                                                                                                                                                                                                                                                    __init__.py:191
           WARNING  connection return: f('admin:admin123@10.73.1.254/command-api', 302, 'Found', <http.client.HTTPMessage object at 0x1062952b0>)                                                                                                                                                                                                      __init__.py:192
           WARNING  Service not running on device 10.73.1.254                                                                                                                                                                                                                                                                                          __init__.py:222
[14:08:42] WARNING  Service not running on device 10.100.164.150 with:                                                                                                                                                                                                                                                                                 __init__.py:191
           WARNING  connection return: ftimed out                                                                                                                                                                                                                                                                                                      __init__.py:192
[14:08:43] WARNING  Service not running on device 10.100.164.150                                                                                                                                                                                                                                                                                       __init__.py:222
           INFO     Inventory .personal/avd-lab.yml loaded                                                                                                                                                                                                                                                                                        check-devices.py:202
           INFO     starting running test on inventory ...                                                                                                                                                                                                                                                                                        check-devices.py:208
           INFO     testing done !                                                                                                                                                                                                                                                                                                                check-devices.py:226
           INFO     retrieve list of result using output_format list                                                                                                                                                                                                                                                                                   __init__.py:106
                 All tests results                  
┏━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Device IP ┃ Test Name ┃ Test Status ┃ Message(s) ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
└───────────┴───────────┴─────────────┴────────────┘
```

### CODE REVIEW

File: anta/inventory/__init__.py [hunk 1]
```diff
@@ -190,8 +187,9 @@ def _is_device_online(self, device: InventoryDevice, timeout: float = 5) -> bool
             setdefaulttimeout(timeout)
```
Comment (mtache) [inline anta/inventory/__init__.py:187]: This line raises a TypeError if `timeout` is not an integer or string (which happened before the use of the `type` arguement in argparse).
If a user does this mistake, I would expect the code to raise an exception.
So I suggest to take the `setdefaulttimeout` function out of the `try` block.
Also, it seems to be tough to cleanly set the timeout for this library (https://stackoverflow.com/questions/372365/set-timeout-for-xmlrpclib-serverproxy), but maybe set back the default timeout to the previous value so it does not affect other sockets opened by the ANTA user would be great (not necessarily `None` like the StackOverflow example but the previous default timeout).

**Review by mtache — DISMISSED** [review]

**Comment by titom73** [comment]
All comments have been adressed in last commit

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+                "-c",
+                "${workspaceFolder}/.personal/ceos-catalog.yml",
+                "--table",
+                "--timeout",
+                "1"
+            ],
+            "console": "integratedTerminal"
+        }
+    ]
+}
```
[the remark this answers is not in the document]
Comment (gmuloc) [reply .vscode/launch.json:27]: in another PR in prep we add some .vscode in the .gitignore though right?

Maybe we can move this to some documentaiton / how-to debug?

File: anta/inventory/__init__.py [hunk 3]
```diff
@@ -494,8 +492,7 @@ def create_device_session(self, host_ip: str) -> bool:
         Returns:
             bool: True if update succeed, False if not
         """
-        logger.debug(
-            f'Searching for device {host_ip} in {[str(dev.host) for dev in self._inventory]}')
+        logger.debug(f'Searching for device {host_ip} in {[str(dev.host) for dev in self._inventory]}')
```
Comment (gmuloc) [inline anta/inventory/__init__.py:495]: So in these next couple of lines we compute again and again some iterators

I would suggest to replace it like this

```suggestion
        logger.debug(f'Searching for device {host_ip} in inventory')
```

and then:

the next few lines to be

```
        device = [dev for dev in self._inventory if str(dev.host) == str(host_ip)][:1] or None
        if device is None:
            return False
        logger.debug(f'Search result is: {device}')
        if device.is_online and not device.established and self._is_ip_exist(host_ip):
            logger.debug(f'Trying to connect to device {str(device.host)}')
            device = self._build_device_session(device=device, timeout=self.timeout)
            # pylint: disable=W0104
            [device if dev.host == device.host else dev for dev in self._inventory]
            return True
```

Though I am not quite sure what this line purpose is:
```
[device if dev.host == device.host else dev for dev in self._inventory]
```
Looks like some list that is never used (not assign to anything) and the method returms True or False

**Review by gmuloc — COMMENTED** [review]
Will need actually run tests and not just comment static code to check the logging stuff. :)

File: anta/inventory/__init__.py [hunk 4]
```diff
@@ -190,8 +187,9 @@ def _is_device_online(self, device: InventoryDevice, timeout: float = 5) -> bool
             setdefaulttimeout(timeout)
             connection.runCmds(1, ['show version'])
         # pylint: disable=W0703
-        except Exception:
-            logger.warning(f'Service not running on device {device.host}')
+        except Exception as exp:
+            logger.warning(f'Service not running on device {device.host} with: ')
+            logger.warning(f'connection return: f{exp}')
```
[the remark this answers is not in the document]
Comment (titom73) [reply anta/inventory/__init__.py:192]: Indeed and wil also solve an issue where logs are not synchronised correctly due to multi-threading (@ksator).

File: anta/inventory/__init__.py (hunk 3 above)
Comment (titom73) [reply anta/inventory/__init__.py:495]: - Regarding first change in the log with inventory, Idea was to give a method to manually check consistence of the search in debug mode. So it is the reason why IP addresses of the inventory are listed. If you think it is useless, we can remove it. (@ksator as we had the exact same discussion internally)
- Other changes are fixed now.

**Review by ksator — COMMENTED** [review]
LGTM

**Review by ksator — APPROVED** [review]

**Pull request merged**